src/view.c

```diff
@@ -1171,6 +1171,18 @@ view_is_tiled(struct view *view)
 		|| view->tiled_region_evacuate);
 }
 
+bool
+view_is_tiled_and_notify_tiled(struct view *view)
+{
+	return
+		((view_is_tiled(view)
+			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_ALWAYS)
+		|| (view->tiled
+			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_EDGE)
+		|| ((view->tiled_region || view->tiled_region_evacuate)
+			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_REGION));
+}
```

Comment:
Just a personal nitpick, I think this is a bit easier to read:
```c
bool
view_is_tiled_and_notify_tiled(struct view *view)
{
	switch (rc.snap_tiling_events_mode) {
	case LAB_TILING_EVENTS_NEVER:
		return false;
	case LAB_TILING_EVENTS_REGION:
		return view->tiled_region || view->tiled_region_evacuate;
	case LAB_TILING_EVENTS_EDGE:
		return view->tiled;
	case LAB_TILING_EVENTS_ALWAYS:
		return view_is_tiled(view);
	}
}
```

It will also complain loudly if we ever end up adding new tiling event states to the enum.

Reply:
How about creating a function like `view_notified_tiled_edges()` which returns the parameter for `wlr_xdg_toplevel_set_tiled()` and share it with `xdg_toplevel_view_notify_tiled()`.

Reply:
> It will also complain loudly if we ever end up adding new tiling event states to the enum.

Thanks, looks indeed easier. That said, the compiler unfortunately still wanted a default case ...

```
../src/view.c:1187:1: warning: control reaches end of non-void function [-Wreturn-type]
 1187 | }
```

Reply:
> How about creating a function like `view_notified_tiled_edges()` which returns the parameter for `wlr_xdg_toplevel_set_tiled()` and share it with `xdg_toplevel_view_notify_tiled()`?

Sorry, I don't understand your suggestion. `xdg_toplevel_view_notify_tiled` is about notifying the client, on the other hand  `view_is_tiled_and_notify_tiled` is about the client being tiled and the client was notified that it is tiled. May be `view_is_tiled_and_notified_tiled` is a better name to avoid this confusion?

Reply:
> > It will also complain loudly if we ever end up adding new tiling event states to the enum.
> 
> Thanks, looks indeed easier. That said, the compiler unfortunately still wanted a default case ...
> 
> ```
> ../src/view.c:1187:1: warning: control reaches end of non-void function [-Wreturn-type]
>  1187 | }
> ```

Compilers can be a bit stupid at times. Just adding a `return false;` after the switch fixes that for me. So there is no `default` case required. I also commented out some enum handler to verify that it still works as intended and got
`../src/view.c:1177:2: error: enumeration value ‘LAB_TILING_EVENTS_ALWAYS’ not handled in switch [-Werror=switch]`.

Reply:
> > How about creating a function like `view_notified_tiled_edges()` which returns the parameter for `wlr_xdg_toplevel_set_tiled()` and share it with `xdg_toplevel_view_notify_tiled()`?
> 
> Sorry, I don't understand your suggestion. `xdg_toplevel_view_notify_tiled` is about notifying the client, on the other hand `view_is_tiled_and_notify_tiled` is about the client being tiled and the client was notified that it is tiled. May be `view_is_tiled_and_notified_tiled` is a better name to avoid this confusion?

I meant something like this:

<details><summary>patch</summary>

```diff
diff --git a/include/view.h b/include/view.h
index b4f90e6a..2bb5fe18 100644
--- a/include/view.h
+++ b/include/view.h
@@ -495,7 +495,7 @@ void view_toggle_always_on_bottom(struct view *view);
 void view_toggle_visible_on_all_workspaces(struct view *view);
 
 bool view_is_tiled(struct view *view);
-bool view_is_tiled_and_notify_tiled(struct view *view);
+uint32_t view_notified_edges(struct view *view);
 bool view_is_floating(struct view *view);
 void view_move_to_workspace(struct view *view, struct workspace *workspace);
 enum ssd_mode view_get_ssd_mode(struct view *view);
diff --git a/src/ssd/ssd-titlebar.c b/src/ssd/ssd-titlebar.c
index 32d6131a..ac771af0 100644
--- a/src/ssd/ssd-titlebar.c
+++ b/src/ssd/ssd-titlebar.c
@@ -123,7 +123,7 @@ ssd_titlebar_create(struct ssd *ssd)
 		set_maximize_alt_icon(ssd, true);
 		ssd->state.was_maximized = true;
 	}
-	if (view_is_tiled_and_notify_tiled(view) && !maximized) {
+	if (view_notified_edges(view) && !maximized) {
 		set_squared_corners(ssd, true);
 		ssd->state.was_tiled_not_maximized = true;
 	}
@@ -193,8 +193,7 @@ ssd_titlebar_update(struct ssd *ssd)
 	struct theme *theme = view->server->theme;
 
 	bool maximized = view->maximized == VIEW_AXIS_BOTH;
-	bool tiled_not_maximized = view_is_tiled_and_notify_tiled(ssd->view)
-		&& !maximized;
+	bool tiled_not_maximized = view_notified_edges(ssd->view) && !maximized;
 
 	if (ssd->state.was_maximized != maximized
 			|| ssd->state.was_tiled_not_maximized != tiled_not_maximized) {
diff --git a/src/view.c b/src/view.c
index eff9b1c0..42f5c632 100644
--- a/src/view.c
+++ b/src/view.c
@@ -1171,16 +1171,49 @@ view_is_tiled(struct view *view)
 		|| view->tiled_region_evacuate);
 }
 
-bool
-view_is_tiled_and_notify_tiled(struct view *view)
-{
-	return
-		((view_is_tiled(view)
-			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_ALWAYS)
-		|| (view->tiled
-			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_EDGE)
-		|| ((view->tiled_region || view->tiled_region_evacuate)
-			&& rc.snap_tiling_events_mode == LAB_TILING_EVENTS_REGION));
+uint32_t
+view_notified_edges(struct view *view)
+{
+	/* Take no action if xdg-shell tiling is disabled */
+	if (rc.snap_tiling_events_mode == LAB_TILING_EVENTS_NEVER) {
+		return 0;
+	}
+
+	enum wlr_edges edge = WLR_EDGE_NONE;
+
+	bool want_edge = rc.snap_tiling_events_mode & LAB_TILING_EVENTS_EDGE;
+	bool want_region = rc.snap_tiling_events_mode & LAB_TILING_EVENTS_REGION;
+
+	/*
+	 * Edge-snapped view are considered tiled on the snapped edge and those
+	 * perpendicular to it.
+	 */
+	if (want_edge) {
+		switch (view->tiled) {
+		case VIEW_EDGE_LEFT:
+			edge = WLR_EDGE_LEFT | WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
+			break;
+		case VIEW_EDGE_RIGHT:
+			edge = WLR_EDGE_RIGHT | WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
+			break;
+		case VIEW_EDGE_UP:
+			edge = WLR_EDGE_TOP | WLR_EDGE_LEFT | WLR_EDGE_RIGHT;
+			break;
+		case VIEW_EDGE_DOWN:
+			edge = WLR_EDGE_BOTTOM | WLR_EDGE_LEFT | WLR_EDGE_RIGHT;
+			break;
+		default:
+			edge = WLR_EDGE_NONE;
+		}
+	}
+
+	if (want_region && view->tiled_region) {
+		/* Region-snapped views are considered tiled on all edges */
+		edge = WLR_EDGE_LEFT | WLR_EDGE_RIGHT |
+			WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
+	}
+
+	return edge;
 }
 
 bool
diff --git a/src/xdg.c b/src/xdg.c
index 2b8bf96e..203a4f42 100644
--- a/src/xdg.c
+++ b/src/xdg.c
@@ -422,46 +422,10 @@ xdg_toplevel_view_set_fullscreen(struct view *view, bool fullscreen)
 static void
 xdg_toplevel_view_notify_tiled(struct view *view)
 {
-	/* Take no action if xdg-shell tiling is disabled */
-	if (rc.snap_tiling_events_mode == LAB_TILING_EVENTS_NEVER) {
-		return;
+	uint32_t edge = view_notified_edges(view);
+	if (edge) {
+		wlr_xdg_toplevel_set_tiled(xdg_toplevel_from_view(view), edge);
 	}
-
-	enum wlr_edges edge = WLR_EDGE_NONE;
-
-	bool want_edge = rc.snap_tiling_events_mode & LAB_TILING_EVENTS_EDGE;
-	bool want_region = rc.snap_tiling_events_mode & LAB_TILING_EVENTS_REGION;
-
-	/*
-	 * Edge-snapped view are considered tiled on the snapped edge and those
-	 * perpendicular to it.
-	 */
-	if (want_edge) {
-		switch (view->tiled) {
-		case VIEW_EDGE_LEFT:
-			edge = WLR_EDGE_LEFT | WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
-			break;
-		case VIEW_EDGE_RIGHT:
-			edge = WLR_EDGE_RIGHT | WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
-			break;
-		case VIEW_EDGE_UP:
-			edge = WLR_EDGE_TOP | WLR_EDGE_LEFT | WLR_EDGE_RIGHT;
-			break;
-		case VIEW_EDGE_DOWN:
-			edge = WLR_EDGE_BOTTOM | WLR_EDGE_LEFT | WLR_EDGE_RIGHT;
-			break;
-		default:
-			edge = WLR_EDGE_NONE;
-		}
-	}
-
-	if (want_region && view->tiled_region) {
-		/* Region-snapped views are considered tiled on all edges */
-		edge = WLR_EDGE_LEFT | WLR_EDGE_RIGHT |
-			WLR_EDGE_TOP | WLR_EDGE_BOTTOM;
-	}
-
-	wlr_xdg_toplevel_set_tiled(xdg_toplevel_from_view(view), edge);
 }
 
 static struct view *

```

</details>

Reply:
> Compilers can be a bit stupid at times. Just adding a `return false;` after the switch fixes that for me. So there is no `default` case required. I also commented out some enum handler to verify that it still works as intended and got `../src/view.c:1177:2: error: enumeration value ‘LAB_TILING_EVENTS_ALWAYS’ not handled in switch [-Werror=switch]`.

ah of course. Adjusted and should be good now.

Reply:
> I meant something like this:

...snip...

> Not a strong preference though.

Thanks, I see. Looks also good and as as far as I see (based on the diff, not tested), the result should be indeed the same. Which version should I go for?